trx_file_memmap/cpp/src/trx.tpp

```diff
@@ -0,0 +1,685 @@
+			}
+
+			new (&(trx->data_per_vertex[base]->_offsets)) Map<Matrix<uint64_t, Dynamic, 1>>(reinterpret_cast<uint64_t *>(trx->streamlines->mmap_off.data()), std::get<0>(shape), std::get<1>(shape));
+			trx->data_per_vertex[base]->_lengths = trx->streamlines->_lengths;
+		}
+
+		else if (folder.compare("dpg") == 0)
+		{
+			std::tuple<int, int> shape;
+			trx->data_per_streamline[base] = new MMappedMatrix<DT>();
+
+			if (size != dim)
+			{
+
+				throw std::invalid_argument("Wrong dpg size/dimensionality");
+			}
+			else
+			{
+				shape = std::make_tuple(1, size);
+			}
+
+			std::string data_name = std::string(basename(const_cast<char *>(base.c_str())));
+			std::string sub_folder = std::string(basename(const_cast<char *>(folder.c_str())));
+
+			trx->data_per_group[sub_folder][data_name] = new MMappedMatrix<DT>();
+			trx->data_per_group[sub_folder][data_name]->mmap = trxmmap::_create_memmap(filename, shape, "r+", ext.substr(1, ext.size() - 1), mem_adress);
+
+			if (ext.compare("float16") == 0)
+			{
+				new (&(trx->data_per_group[sub_folder][data_name]->_matrix)) Map<Matrix<half, Dynamic, Dynamic>>(reinterpret_cast<half *>(trx->data_per_group[sub_folder][data_name]->mmap.data()), std::get<0>(shape), std::get<1>(shape));
+			}
+			else if (ext.compare("float32") == 0)
+			{
+				new (&(trx->data_per_group[sub_folder][data_name]->_matrix)) Map<Matrix<float, Dynamic, Dynamic>>(reinterpret_cast<float *>(trx->data_per_group[sub_folder][data_name]->mmap.data()), std::get<0>(shape), std::get<1>(shape));
+			}
+			else
+			{
+				new (&(trx->data_per_group[sub_folder][data_name]->_matrix)) Map<Matrix<double, Dynamic, Dynamic>>(reinterpret_cast<double *>(trx->data_per_group[sub_folder][data_name]->mmap.data()), std::get<0>(shape), std::get<1>(shape));
+			}
+		}
+
+		else if (folder.compare("groups") == 0)
+		{
+			std::tuple<int, int> shape;
+			if (dim != 1)
+			{
+				throw std::invalid_argument("Wrong group dimensionality");
+			}
+			else
+			{
+				shape = std::make_tuple(size, 1);
+			}
+			trx->groups[base] = new MMappedMatrix<uint32_t>();
+			trx->groups[base]->mmap = trxmmap::_create_memmap(filename, shape, "r+", ext.substr(1, ext.size() - 1), mem_adress);
+			new (&(trx->groups[base]->_matrix)) Map<Matrix<uint32_t, Dynamic, Dynamic>>(reinterpret_cast<uint32_t *>(trx->groups[base]->mmap.data()), std::get<0>(shape), std::get<1>(shape));
+		}
+		else
+		{
+			// TODO: logger
+			std::cout << elem_filename << " is not part of a valid structure." << std::endl;
+		}
+	}
+	if (trx->streamlines->_data.size() == 0 || trx->streamlines->_offsets.size() == 0)
+	{
+
+		throw std::invalid_argument("Missing essential data.");
+	}
+
+	return trx;
+}
+
+template <typename DT>
+TrxFile<DT> *load_from_zip(std::string filename)
+{
+	// TODO: check error values
+	int *errorp;
+	zip_t *zf = zip_open(filename.c_str(), 0, errorp);
+	json header = load_header(zf);
+
+	std::map<std::string, std::tuple<int, int>> file_pointer_size;
```

Comment:
The file size/offset may need to be greater than `int`, I am having a problem here loading files >2GB.  Afaik they are positive, so `uint` gives a bit more of a range, but `long long` may cover other larger cases. Changing it (and downstream in fn calls) to `long long` worked here.

Reply:
Sounds good! Would you like to add your updates?